rfcs/2020-08-31-3645-graphql-api.md

```diff
@@ -0,0 +1,279 @@
+
+This is largely TBD, but the basic mechanisms are there to allow for flexible auth when we need it.
+
+## Doc-level Proposal
+
+A Vector observability layer has already been agreed internally. Work is underway.
+
+This proposal discusses the protocol which will govern communications between a running Vector instance, and a web UI and CLI.
+
+The result of this proposal won't directly impact user interaction with observability tooling.
+
+## Rationale
+
+We have built a decent body of experience with GraphQL at Timber, albeit on non-Vector projects.
+
+The tooling I am proposing here represents the same stack (save for swapping Apollo for Urql.)
+
+In a previous Timber project, we used [gqlgen](https://github.com/99designs/gqlgen) as our server library, written in Go. The tooling with async-graphql is similar, though Rust's language features enable a more composible 'code first' approach using macros.
+
+I believe the type of data we are consuming benefits from a strongly typed interface, and that compile-time client generation will significantly reduce the time-to-market of what will be already be a very complex front-end web app.
+
+## Prior Art
+
+See above
+
+## Drawbacks
+
+* GraphQL is generally geared toward graph data (hence the name), although this isn't a hard requirement. IMO, it's a solid general-purpose API interface that mimics what we'd generally wind up creating internally anyway.
+* Queries with unknown levels of nesting can't be queried statically; generally, a need for queries with flatter structures arise.
+* Versioning is trickier, given the lack of distinct endpoints. Nothing stopping multiple endpoints representing versions, although typically this is done in the schema.
+* Some annoying rough edges -- the lack of input unions, for example, typically worked around with more specific queries.
+* Reliance on a relatively young library rather than a 'pure' client like a stdlib tcp server.
+* A need to invest, longer-term, in understanding and contributing to an external library.
+
+## Alternatives
+
+Essentially any browser-compatible protocol could be used, and any text/binary format. This could include other libs such as Protobuf, which might form of a hybrid of the aforementioned approaches.
+
+The maturity of tooling would need further investigation, as well as the development experience of working with any given format.
+
+## Outstanding Questions
+
+* What are the performance characteristics we care about?
+* How many clients do we assume we will be serving across the total of all `vector top` and Vector UI requests?
+* Is JSON an acceptable format for data exchange?
+* What's the GraphQL Client tooling like in Rust, to enable UI-like comms with `vector top`?
```

Comment:
This is a big one for me. Obviously the browser-based frontend is a priority, but it's not the only client we'll want to make sure is robust and fully-featured.

Reply:
If the client can send JSON payloads over HTTP, then it can handle GraphQL queries and mutations. If it can connect HTTP and handle WS upgraded and longer lived TCP connections, it can handle streaming. If it has a dedicated GraphQL client that can do those things _and_ add semantic meaning to the data it's sending and receiving -- then it can benefit from the type hinting espoused in the [development workflow video](https://www.loom.com/share/9ca38feb21bb488f92d729df0148e029).

Anything else we need, we can either bolt on, or work around as an aside. Namely:

- Compression. Add `permessage-deflate` or a DIY compression via `Sec-WebSocket-Extensions`.
- Message serialization. Transform JSON to binary or something else.
- gRPC. Have a separate endpoint for it to work alongside. Doesn't have to be either/or.

For `vector top` specifically, there's https://github.com/graphql-rust/graphql-client. I haven't dug into it much yet, but evidently it has some form of subscription support and even works as WASM in the browser. There's probably others-- I've only looked superficially but will dive in more once _how_ we're observing topology is solved.

Reply:
Wondering if we're (possibly) talking about having different types of transports supported for different types of clients (eventually).   Not sure what that would mean implementation wise but I don't think GraphQL itself is dictating how the transport of messages occurs.  Usually, when you're talking about GraphQL you're talking about doing stuff over HTTP between a client (browser) and a server.  That's the most common use case and really the only one I have any experience with.  I went back today was watching this video from Lee Byron (one of the main people that developed the GraphQL spec) where he introduced subscriptions.  This is a much more abstract idea about events streams that is not necessarily tied to any particular transport method.  He talks about using messaging pub-sub type protocols like MQTT (or ZeroMQ) inside of FB and even web hook callbacks.  

If you've never watched this video, it may be worth taking a look...
https://www.youtube.com/watch?v=bn8qsi8jVew

Reply:
> For `vector top` specifically, there's https://github.com/graphql-rust/graphql-client. I haven't dug into it much yet, but evidently it has some form of subscription support and even works as WASM in the browser.

Yeah, this is something that'd be nice to dig into a bit. I'm sure we can flesh out support ourselves if needed, but it's going to be important to have a first-class, ergonomic, performant Rust interface to this API.

> Wondering if we're (possibly) talking about having different types of transports supported for different types of clients (eventually).

That's exactly what I'm curious about. If we're using a non-browser client it could be nice to swap out WebSockets for plain old TCP, but it may not really matter.